### PR TITLE
test(util): add unit tests for pkg/util/hash

### PR DESCRIPTION
**Pull request by goyalpalak18**
**What type of PR is this?**
/kind cleanup

---

### Description
I noticed that the `pkg/util/hash` package currently has **0% test coverage**.

This seemed risky because `DeepHashObject` is a foundational function used to generate deterministic names for resources (like PolicyNames). If this function were to behave inconsistently across Go versions or library updates, it could silently break resource propagation without us knowing.

---

### The Fix
I added a comprehensive `hash_test.go` file using the standard table-driven test pattern found elsewhere in the codebase.

Instead of just checking basic success, I ensured we cover the edge cases that matter for stability:
* **Nil handling:** Ensuring it doesn't panic on nil values.
* **Pointers vs Values:** Verifying they hash consistently.
* **State Safety:** confirming the hasher resets correctly between calls so state doesn't "leak" from one object to the next.

This brings the statement coverage for the package to **100%**.

---

### How to Test
1.  Navigate to the package directory.
2.  Run the tests with coverage enabled:

```bash
go test -v -cover ./pkg/util/hash/...

### CODE REVIEW

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks @goyalpalak18 
It can be seen that the content of this test is quite comprehensive. However, on the other hand, I feel it is a bit complex. Using so much code to cover the code under test might add some maintenance burden compared to the code being tested. From this perspective, the cost might be a bit too high.

**Comment by goyalpalak18** [comment]
Thanks for the review @XiShanYongYe-Chang!

I totally get where you're coming from—it does look like a lot of test code for such a small function. 

My main motivation was that since `DeepHashObject` drives resource naming (like `GeneratePolicyName`), stability is super critical here. If the underlying library (`kr/pretty`) changes or we hit a weird edge case, the hashes could silently shift and break resource propagation. I wanted these tests to act as a strict safety net against that.

That said, I'm happy to trim it down if you feel the maintenance cost is too high. Just let me know if you'd prefer I drop the complex nested cases and stick to the basics!

**Comment by XiShanYongYe-Chang** [comment]
> If the underlying library (kr/pretty) changes or we hit a weird edge case

Your concern is valid, but I think such drastic changes almost impossible, so I suggest we retain an appropriate level of testing to avoid adding to our maintenance burden.

**Comment by goyalpalak18** [comment]
@XiShanYongYe-Chang Done. I cut it down to just the 3 main test cases to keep it clean. Thanks for the feedback!

File: pkg/util/hash/hash_test.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2021 The Karmada Authors.
```
Comment (XiShanYongYe-Chang) [inline pkg/util/hash/hash_test.go]: 2021 -> 2026

**Comment by XiShanYongYe-Chang** [comment]
Thanks
/assign
/gemini review

File: pkg/util/hash/hash_test.go (hunk 1 above)
Comment (goyalpalak18) [reply pkg/util/hash/hash_test.go]: I've updated the copyright year to 2025. Thanks!

Comment (XiShanYongYe-Chang) [reply pkg/util/hash/hash_test.go]: Sorry, I replied incorrectly; it's already 2026 now.

**Comment by XiShanYongYe-Chang** [comment]
Hi @goyalpalak18, can you help sign off on the commits like before?

**Comment by goyalpalak18** [comment]
> Hi @goyalpalak18, can you help sign off on the commits like before?

Done! I have signed off on the commit.

File: pkg/util/hash/hash_test.go (hunk 1 above)
Comment (XiShanYongYe-Chang) [reply pkg/util/hash/hash_test.go]: @goyalpalak18 This one seems to have been missed.

Comment (goyalpalak18) [reply pkg/util/hash/hash_test.go]: @XiShanYongYe-Chang Fixed the copyright year to 2026.

**Comment by XiShanYongYe-Chang** [comment]
Thanks
lgtm
Can you help squash the commits into one?

**Comment by goyalpalak18** [comment]
> Thanks lgtm Can you help squash the commits into one?

@XiShanYongYe-Chang I've squashed the commits into one.
